Deglobalize repair history maps

bhalevy opened the pull request:
Change a8ad385ecd3 introduced
```
thread_local std::unordered_map<utils::UUID, seastar::lw_shared_ptr<repair_history_map>> repair_history_maps;
```

We're trying to avoid global scoped variables as much as we can so this should probably be embedded in some sharded service.

This series moves the thread-local `repair_history_maps` instances to `compaction_manager`
and passes a reference to the shard compaction_manager to functions that need it for compact_for_query
and compact_for_compaction.

Since some paths don't need it and don't have access to the compactio_manager,
the series introduced `utils::optional_reference<T>` that allows to pass nullopt.
In this case, `get_gc_before_for_key` behaves in `tombstone_gc_mode::repair` as if the table wasn't repaired and tombstones are not garbage-collected.

Fixes #11208

psarna commented on repair/row_level.cc:
```diff
@@ -3008,14 +3008,14 @@ future<> repair_service::load_history() {
             rlogger.debug("Loading repair history for keyspace={}, table={}, table_uuid={}, repair_time={}, range={}",
                     entry.ks, entry.cf, entry.table_uuid, entry.ts, range);
-            co_await get_db().invoke_on_all([&entry, range, repair_time] (replica::database& local_db) {
-                try {
```
Isn't it ever useful to know the context (== shard here) on which the operation failed? We lose this context after this change.

bhalevy replied:
I don't think so.  @asias what do you think?

asias replied:
I think the change here is ok. If we want to know which shard fails, we can log it inside get_compaction_manager().update_repair_time().

psarna commented on repair/row_level.cc:
```diff
@@ -3009,17 +3008,13 @@ future<> repair_service::load_history() {
             rlogger.debug("Loading repair history for keyspace={}, table={}, table_uuid={}, repair_time={}, range={}",
                     entry.ks, entry.cf, entry.table_uuid, entry.ts, range);
-            co_await get_db().invoke_on_all([entry, range, repair_time] (replica::database& local_db) -> future<> {
+            co_await get_db().invoke_on_all([&entry, range, repair_time] (replica::database& local_db) {
```
nitpick: it might be a little more idiomatic to just capture the needed value here instead of a reference: `table_uuid = entry.table_uuid`

bhalevy replied:
yes, it's possible

bhalevy replied:
(in the next patch... this one still needs the reference to entry for printing the warning)

psarna commented on test/lib/sstable_test_env.hh:
```diff
@@ -91,6 +95,11 @@ public:
 
+
+    compaction_manager& get_compaction_manager() noexcept {
+        return *_compaction_manager;
```
Wait, wasn't the definition of this attribute `compaction_manager _compaction manager`? In this case this code should not compile

bhalevy replied:
oops, the method is not used until 3628e26e86 where it's fixed.
I can remove it here entirely until the patch that introduces its users.

psarna commented on utils/optional_reference.hh:
```diff
@@ -0,0 +1,67 @@
+namespace utils {
+
+template <class T>
+class optional_reference {
```
kind of like https://en.cppreference.com/w/cpp/experimental/observer_ptr?

bhalevy replied:
Hmm, yeah, seems functionally equivalent.
I'll try to use that instead then.

denesb replied:
We already have a native language construct for this: a pointer. Why reinvent it?

bhalevy replied:
The main reason is the ability to initialize with a reference that's guaranteed not to be null, or with nullopt otherwise.
Also, since it's not a pointer, it can't be modified via pointer arithmetic and so on.

denesb replied:
You can do all this with a pointer. If you make your pointer `const T* const` you have all that you list above: a pointer that can be initialized from a reference or `nullptr` and cannot be modified via arithmetic and so on.

bhalevy replied:
It looks like `observer_ptr` provides that safe abstraction in a clear way.
Why not use it? Since it's experimental?

psarna replied:
@bhalevy I actually posted the link as "fun fact", it's probably safer to avoid experimental unless absolutely necessary. I actually like Botond's idea of a pointer with consts put in every possible corner.

bhalevy replied:
ok, if you guys insist...

denesb replied:
I hope you are trolling :grin:. Why do we want to suddenly replace good ol' language constructs with template classes?
I honestly very much dislike `std::reference_wrapper<T>` too, and think it only has place for covering a niche of binding references bind markers.

xemul replied:
> I honestly very much dislike std::reference_wrapper<T> too, and think it only has place for covering a niche of binding references bind markers.

_Just_ `reference_wrapper`? It's all over the `sharded<T>::start()` stuff! Is it this "binding references bind markers" case you mentioned?

denesb replied:
Yes, for example.

asias replied:
I am not a fan of the home-grown optional_reference neither.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1925/

bhalevy commented:
In v2 ([2dc786d](https://github.com/scylladb/scylladb/commit/2dc786da6077b48409bbec70d93da39c4867397c)):
- using `compaction_manager_opt = const compaction_manager* const`
- rebased and cleaned up.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/1970/

denesb reviewed (commented):
I dislike having to pass `compaction_manager` to every place which wishes to do compaction. Wouldn't it be better to narrow this down to just the repair history map? Do you anticipate other things we might want to obtain from the compaction manager?

bhalevy commented:
> I dislike having to pass `compaction_manager` to every place which wishes to do compaction. Wouldn't it be better to narrow this down to just the repair history map? Do you anticipate other things we might want to obtain from the compaction manager?

Not really.  I don't anticipate other things we need from the compaction manger at the moment.
However, I'm not sure that being too specific and passing down only (optional) repair map is the right thing todo.
The mid-layers don't care about the mechanism, they just need to compact the results they retrieve somehow, so I personally see no problem in involving the compaction_manager in the process.  Otherwise, if/when we change the tombstone gc logic we might need to change this to something else.

Maybe instead of passing a pointer to the compaction manager we need some abstraction holding stuff for compacting mutations and the likes, and that can hold a point to the repair maps today and maybe some more information in the future if and when needed.  Though just getting it from the compaction manager doesn't seem unreasonable to me either :)

bhalevy commented:
In v3 ([bf28a20](https://github.com/scylladb/scylladb/commit/bf28a20aa297903d70a6135c70641ff4181002b8)):
- view: pass base table to view_update_builder
  - instead of passing a ref to compaction_manager
- tombstone_gc: introduce tombstone_gc_state
  - does most of the transition by defining `class tombstone_gc_state`
  - that keeps a pointer to the repair history maps
  - and provides the interface to them, instead of the free functions in tombstone_gc.{cc,hh}
- tombstone_gc: deglobalize repair_history_maps
  - became much small consequently
  - only moves the maps from thread local scope to members of compaction_manager
- dropped "tombstone_gc: move methods away to compaction_manager"
  - not needed anymore since `tombstone_gc_state` implementation is kept in tombstone_gc.cc

denesb commented on test/lib/sstable_test_env.hh:
```diff
@@ -14,6 +14,8 @@
 #include <seastar/util/defer.hh>
 
 #include "sstables/sstables.hh"
+#include "compaction/compaction_manager.hh"
```
nit: commitlog says "start a compaction..." but I don't see anything started here, just created.

bhalevy replied:
It's a degenerate start since it's a single shard that's constructed with `compaction_manager(compaction_manager::for_testing_tag{})`.
I'll reword the commit message to make it clearer.

denesb commented on service/storage_proxy.cc:
```diff
@@ -4093,12 +4097,12 @@ class data_read_resolver : public abstract_read_resolver {
         auto&& ranges = cmd.slice.row_ranges(s, m.key());
         bool always_return_static_content = cmd.slice.options.contains<query::partition_slice::option::always_return_static_content>();
-        mp.compact_for_query(s, m.decorated_key(), cmd.timestamp, ranges, always_return_static_content, is_reversed, limit);
+        mp.compact_for_query(s, m.decorated_key(), cmd.timestamp, ranges, always_return_static_content, is_reversed, limit, gc_state);
```
I'm not sure we need the gc state here. Replicas should only send non-purgeable tombstones already, so we can expect to not have to actually purge any tombstones here.

bhalevy replied:
okay, I'll remove it and retest.

denesb reviewed (changes requested):
LGTM, I like this version much more.
I think we don't need the storage proxy changes -- reconciliation should not have any purgeable tombstones already. These changes do no harm of course, other than the added noise so if unnecessary maybe we should avoid it.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2095/

bhalevy commented:
In v4 ([4d92350](https://github.com/scylladb/scylladb/commit/4d92350bbaae49ec29594835d4acc18e922d9b88))
- test: sstables test_env: hold a compaction manager member for testing
  - clarified commit message
- tombstone_gc: introduce tombstone_gc_state
  - dropped changes to storage_proxy
  - compact_for_query does not get a gc_state param, but rather uses a null tombstone_gc_state,

avikivity commented on test/lib/sstable_test_env.hh:
```diff
@@ -33,10 +35,12 @@ class test_env {
         cache_tracker cache_tracker;
         test_env_sstables_manager mgr;
         reader_concurrency_semaphore semaphore;
+        compaction_manager compaction_manager;
```
Does it need to be stopped?

This expanding the scope of test_env (from testing single sstables to testing compaction), maybe we need a different test_env for that.

avikivity commented on compaction/compaction_strategy.cc:
```diff
@@ -86,7 +86,9 @@ std::optional<sstring> compaction_strategy_impl::get_value(const std::map<sstrin
 
-compaction_strategy_impl::compaction_strategy_impl(const std::map<sstring, sstring>& options) {
+compaction_strategy_impl::compaction_strategy_impl(compaction_manager& cm, const std::map<sstring, sstring>& options)
+    : _compaction_manager(&cm)
```
This is entangling things too much together. The compaction strategy is used by compaction_manager to decide what to compact, and now the dependency is circular. It's making everything more difficult.

avikivity replied:
Yes, but the strategy is a leaf (at least relatively) on the dependency graph now. compaction_manager is near the root. Let's not make it circular with such large span.

avikivity commented on compaction/compaction_manager.hh:
```diff
@@ -8,6 +8,9 @@
+#include <boost/icl/interval.hpp>
+#include <boost/icl/interval_map.hpp>
+
 #include <seastar/core/semaphore.hh>
```
Why this? We

bhalevy replied:
```
class repair_history_map {
public:
    boost::icl::interval_map<dht::token, gc_clock::time_point, boost::icl::partial_absorber, std::less, boost::icl::inplace_max> map;
};
```

This definition was moved to compaction_manager.hh in this patch.

raphaelsc commented on compaction/compaction.cc:
```diff
@@ -1766,7 +1770,7 @@ get_fully_expired_sstables(const table_state& table_s, const std::vector<sstable
     for (auto& sstable : overlapping) {
-        auto gc_before = sstable->get_gc_before_for_fully_expire(compaction_time);
+        auto gc_before = sstable->get_gc_before_for_fully_expire(compaction_time, cm.get_tombstone_gc_state());
         if (sstable->get_max_local_deletion_time() >= gc_before) {
```
Why don't we get tombstone_gc_state using _table_s.get_compaction_strategy().get_tombstone_gc_state();?

I think we shouldn't pass compaction_manager around, it's like the highest layer in compaction layer and should never be passed down to any lower level entity.

bhalevy replied:
I'll give it a shot

raphaelsc commented on compaction/leveled_compaction_strategy.cc:
```diff
@@ -47,8 +47,8 @@ compaction_descriptor leveled_compaction_strategy::get_sstables_for_compaction(t
-            auto gc_before1 = i->get_gc_before_for_drop_estimation(compaction_time);
-            auto gc_before2 = j->get_gc_before_for_drop_estimation(compaction_time);
+            auto gc_before1 = i->get_gc_before_for_drop_estimation(compaction_time, get_tombstone_gc_state());
+            auto gc_before2 = j->get_gc_before_for_drop_estimation(compaction_time, get_tombstone_gc_state());
```
If you want to avoid the burden of passing tombstone_gc_state() in some places, you can define table_state::tombstone_gc_state(), which can in turn retrieve it from table's compaction manager. Then unit tests can have a table state implemented which retrieves the state that can be stored anywhere.

avikivity replied:
Absolutely, table_state was designed for this.

bhalevy replied:
this is simpler and more elegant indeed.

raphaelsc commented on db/view/view.hh:
```diff
@@ -206,6 +206,10 @@ private:
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-private-field"
+    const replica::table& _base;
+#pragma GCC diagnostic pop
```
Ok, saw you removed it later.

bhalevy replied:
right, it's only temporary

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2100/

bhalevy commented:
In v5 ([2e78989](https://github.com/scylladb/scylladb/commit/2e789891d98e12f6101259bd3084ed3c91dc43ba)):
- rebased, cleaned up, and retested
- dropped:
  - test: sstables test_env: hold a compaction manager member for testing
  - compaction_strategy: get a reference to compaction_manager when constructed
- use table_state to get to the tombstone_gc_state
  - instead of passing reference to compaction manager
- tombstone_gc: introduce tombstone_gc_state
  - greatly simplified and distilled to the minimum
  - the rest was broken down to a series of individual changes:
    - replica: table: add get_compaction_manager function
    - compaction_manager: add tombstone_gc_state
    - replica: database: get_tombstone_gc_state from compaction_manager
    - repair: row_level: repair_update_system_table_handler: get get_tombstone_gc_state for db compaction_manager
    - db: view: get_tombstone_gc_state from compaction_manager
    - compaction: table_state: add virtual get_tombstone_gc_state method
    - sstables: get_gc_before_*: get tombstone_gc_state from caller
    - mutation_readers: pass tombstone_gc_state to compating_reader
    - mutation_partition: compact_for_compaction: get tombstone_gc_state
    - mutation_partition: compact_for_compaction_v2: get tombstone_gc_state
    - mutation_compactor: pass tombstone_gc_state to compact_mutation_state

scylladb-promoter commented:
CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2109/

raphaelsc commented:
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2109/

we can probably forward declare tombstone_gc_state in readers/compacting.hh to avoid the error

raphaelsc commented:
series look good to me

bhalevy commented:
In v6 ([e9cfe9e](https://github.com/scylladb/scylladb/commit/e9cfe9e572a005c46d2c98f99443655a3e7fea99)):
- added forward declaration of `tombstone_gc_state` in readers/compacting.hh to keep the header file self-sufficient.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/releng/job/Scylla-CI/2111/

bhalevy commented:
@denesb ping, can this PR be merged today?

denesb commented:
> @denesb ping, can this PR be merged today?

Queued.
